miner/tx_ringbuffer.go

```diff
@@ -0,0 +1,65 @@
+package miner
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// TxRingBuffer is a ring buffer of transaction hashes, it has one method to add
+// transaction hashes and another method to check if a transaction hash is in the buffer.
+type TxRingBuffer struct {
+	ring  []common.Hash
+	m     map[common.Hash]struct{}
+	index int
+	mux   *sync.RWMutex
+}
+
+// NewTxRingBuffer creates a new transaction ring buffer with the given size.
+func NewTxRingBuffer(size int) *TxRingBuffer {
+	return &TxRingBuffer{
+		ring: make([]common.Hash, size),
+		m:    make(map[common.Hash]struct{}),
+		mux:  &sync.RWMutex{},
+	}
+}
+
+// Add adds a new transaction hash to the ring buffer, if the buffer is
+// at capacity, the oldest transaction hash is overwritten.
+func (b *TxRingBuffer) Add(tx common.Hash) {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	if _, exists := b.m[tx]; exists == true {
+		return
+	}
+
+	// Clean up old transaction hash from map
+	oldHash := b.ring[b.index]
```

Comment:
wonder what happens when the ring is not yet full... this returns zero and delete(zero) doesn't break?

Reply:
It will be the empty hash, and delete the empty hash which will be a no-op.